sdk/keyvault/test-resources-cleanup.py

```diff
@@ -31,25 +32,60 @@
-]
-for certificate in deleted_test_certificates:
-    cert_client.purge_deleted_certificate(certificate.name)
-
-test_keys = [k for k in key_client.list_properties_of_keys() if k.name.startswith("livekvtest")]
-for key in test_keys:
-    key_client.begin_delete_key(key.name).wait()
-deleted_test_keys = [k for k in key_client.list_deleted_keys() if k.name.startswith("livekvtest")]
-for key in deleted_test_keys:
-    key_client.purge_deleted_key(key.name)
-
-test_secrets = [s for s in secret_client.list_properties_of_secrets() if s.name.startswith("livekvtest")]
-for secret in test_secrets:
-    secret_client.begin_delete_secret(secret.name).wait()
-deleted_test_secrets = [s for s in secret_client.list_deleted_secrets() if s.name.startswith("livekvtest")]
-for secret in deleted_test_secrets:
-    secret_client.purge_deleted_secret(secret.name)
+async def delete_certificates():
+    coroutines = []
+
+    test_certificates = cert_client.list_properties_of_certificates()
+    async for certificate in test_certificates:
+        if certificate.name.startswith("livekvtest"):
+            coroutines.append(cert_client.delete_certificate(certificate.name))
+
+    return await asyncio.gather(*coroutines)
+
+async def delete_keys_and_secrets():
+    coroutines = []
+
+    test_keys = key_client.list_properties_of_keys()
+    async for key in test_keys:
+        if key.name.startswith("livekvtest"):
+            coroutines.append(key_client.delete_key(key.name))
+
+    test_secrets = secret_client.list_properties_of_secrets()
+    async for secret in test_secrets:
+        if secret.name.startswith("livekvtest"):
+            coroutines.append(secret_client.delete_secret(secret.name))
+
+    return await asyncio.gather(*coroutines)
+
+async def purge_resources():
+    coroutines = []
+
+    deleted_test_certificates = cert_client.list_deleted_certificates(include_pending=True)
+    async for certificate in deleted_test_certificates:
+        if certificate.name.startswith("livekvtest"):
+            coroutines.append(cert_client.purge_deleted_certificate(certificate.name))
+
+    deleted_test_keys = key_client.list_deleted_keys()
+    async for key in deleted_test_keys:
+        if key.name.startswith("livekvtest"):
+            coroutines.append(key_client.purge_deleted_key(key.name))
+
+    deleted_test_secrets = secret_client.list_deleted_secrets()
+    async for secret in deleted_test_secrets:
+        if secret.name.startswith("livekvtest"):
+            coroutines.append(secret_client.purge_deleted_secret(secret.name))
+
+    return await asyncio.gather(*coroutines)
+
+async def close_sessions():
+    await credential.close()
+    await cert_client.close()
+    await key_client.close()
+    await secret_client.close()
+
+loop = asyncio.get_event_loop()
+loop.run_until_complete(delete_certificates())
+loop.run_until_complete(delete_keys_and_secrets())
+loop.run_until_complete(purge_resources())
```

Comment:
The script might be faster if it scheduled all these coroutines at once.

Reply:
It might be, but Key Vault raises an error when a coroutine is scheduled to delete a key or secret that's created for a certificate. Do you know of any way around that?

Reply:
Keys and secrets have a "managed" property to indicate whether they were so created (which we don't expose for secrets! #17129 😨). You could skip deleting and purging the managed ones, assuming Key Vault deletes/purges them with their associated certificate (naively seems like it must do this but I haven't checked). I expect the decrease in total requests would have a much greater impact than scheduling all the coroutines at once. Anyway, it's not something you need to investigate, unless you're too curious 😸